source/opt/dominator_tree.cpp

```diff
@@ -0,0 +1,395 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dominator_tree.h"
+#include "cfa.h"
+
+#include <iostream>
+#include <set>
+
+using namespace spvtools;
+using namespace spvtools::opt;
+
+namespace {
+
+// Wrapper around CFA::DepthFirstTraversal to provide an interface to perform
+// depth first search on generic BasicBlock types. Will call post and pre order
+// user defined functions during traversal
+//
+// BBType - BasicBlock type. Will either be ir::BasicBlock or DominatorTreeNode
+// SuccessorLambda - Lamdba matching the signature of 'const
+// std::vector<BBType>*(const BBType *A)'. Will return a vector of the nodes
+// succeding BasicBlock A.
+// PostLambda - Lamdba matching the signature of 'void (const BBType*)' will be
+// called on each node traversed AFTER their children.
+// PreLambda - Lamdba matching the signature of 'void (const BBType*)' will be
+// called on each node traversed BEFORE their children.
+template <typename BBType, typename SuccessorLambda, typename PreLambda,
+          typename PostLambda>
+static void depthFirstSearch(const BBType* BB, SuccessorLambda successors,
+                             PreLambda pre, PostLambda post) {
+  // Ignore backedge operation.
+  auto nop_backedge = [](const BBType*, const BBType*) {};
+  CFA<BBType>::DepthFirstTraversal(BB, successors, pre, post, nop_backedge);
+}
+
+// Wrapper around CFA::DepthFirstTraversal to provide an interface to perform
+// depth first search on generic BasicBlock types. This overload is for only
+// performing user defined post order.
+//
+// BBType - BasicBlock type. Will either be ir::BasicBlock or DominatorTreeNode
+// SuccessorLambda - Lamdba matching the signature of 'const
+// std::vector<BBType>*(const BBType *A)'. Will return a vector of the nodes
+// succeding BasicBlock A.
+// PostLambda - Lamdba matching the signature of 'void (const BBType*)' will be
+// called on each node traversed after their children.
+template <typename BBType, typename SuccessorLambda, typename PostLambda>
+static void depthFirstSearchPostOrder(const BBType* BB,
+                                      SuccessorLambda successors,
+                                      PostLambda post) {
+  // Ignore preorder operation.
+  auto nop_preorder = [](const BBType*) {};
+  depthFirstSearch(BB, successors, nop_preorder, post);
+}
+
+// Small type trait to get the function class type.
+template <typename BBType>
+struct GetFunctionClass {
+  using FunctionType = ir::Function;
+};
+
+// This helper class is basically a massive workaround for the current way that
+// depth first is implemented.
```

Comment:
The current DFT function takes in an std::function to return the predecessors and one to return the successors of any given BB. They each return a 'const std::vector<BB*>*' which means (as it's a pointer) we have to precompute the two lists and maintain them. For our case it could just be done on the fly since they aren't used again.

Workaround is probably not the right word, I'll update the comment to make it clearer why we need that class.

Reply:
Ah, thanks.  Now I understand the issue better.  I'm fine with this.